AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/TgzPackageExtract.kt

```diff
@@ -0,0 +1,330 @@
+     *
+     * @param tarballFile the .tgz file to extract
+     * @param addonsDir   the AnkiDroid addons directory
+     * @return the temp directory.
+     * @throws FileNotFoundException if .tgz file or ungzipped file i.e. .tar file not found
+     * @throws IOException
+     */
+    @Throws(Exception::class)
+    fun extractTarGzipToAddonFolder(tarballFile: File, addonsDir: File) {
+
+        require(isGzip(tarballFile)) { context.getString(R.string.not_valid_js_addon, tarballFile.absolutePath) }
+
+        if (!addonsDir.exists()) {
+            context.getString(R.string.could_not_create_dir, addonsDir.absolutePath)
+        }
+
+        // Make sure we have 2x the tar file size in free space (1x for tar file, 1x for unarchived tar file contents
+        requiredMinSpace = tarballFile.length() * 2
+        availableSpace = Utils.determineBytesAvailable(addonsDir.canonicalPath)
+        if (requiredMinSpace > availableSpace) {
+            Timber.e("Not enough space to extract file, need %d, available %d", requiredMinSpace, availableSpace)
+            throw IOException(context.getString(R.string.import_log_insufficient_space_error, Formatter.formatFileSize(context, requiredMinSpace), Formatter.formatFileSize(context, availableSpace)))
+        }
+
+        // If space available then unGZip it
+        val tarTempFile = unGzip(tarballFile, addonsDir)
+        tarTempFile.deleteOnExit()
+
+        // Make sure we have sufficient free space
+        val unTarSize = calculateUnTarSize(tarTempFile)
+        if (unTarSize > availableSpace) {
+            Timber.e("Not enough space to untar, need %d, available %d", unTarSize, availableSpace)
+            throw IOException(context.getString(R.string.import_log_insufficient_space_error, Formatter.formatFileSize(context, unTarSize), Formatter.formatFileSize(context, availableSpace)))
+        }
+
+        try {
+            // If space available then unTar it
+            unTar(tarTempFile, addonsDir)
+        } catch (e: IOException) {
+            Timber.w("Failed to unTar file")
+        }
+    }
+
+    /**
+     * UnGZip a file: a .tgz file will become a .tar file.
+     *
+     * @param inputFile The [File] to ungzip
+     * @param outputDir The directory where to put the ungzipped file.
+     * @return a [File] pointing to the ungzipped file.
+     * @throws FileNotFoundException
+     * @throws IOException
+     */
+    @Throws(FileNotFoundException::class, IOException::class)
+    fun unGzip(inputFile: File, outputDir: File): File {
+        Timber.i("Ungzipping %s to dir %s.", inputFile.absolutePath, outputDir.absolutePath)
+
+        // rename '-4' to remove the '.tgz' extension and add .tar extension
+        val outputFile = File(outputDir, inputFile.name.substring(0, inputFile.name.length - 4) + ".tar")
+
+        count = 0
+        total = 0
+        data = ByteArray(BUFFER)
+
+        try {
+            // file being unGzipped must not be greater than 100MB
```

Comment:
You are still not taking in to account that *during ungzip you could have less than the space needed for the tar file or the unpacking of tar file. If at any time the count of bytes ungzipped is more than .5x space available at start, you must throw.

```suggestion
            // file being unGzipped must not be greater than TOO_BIG_SIZE or half of available space
```